File: Problem 7.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : needed to refer to lecture to see how to double the search space but remembered the logic

class Solution {
    public int search(ArrayReader reader, int target) {
        
        int low = 0;
        int high = 1;
        
        while(reader.get(high) < target){
            low = high;
            high = 2 * high;
        }
        
        return binarySearch(reader, low, high, target);
    }
    
    private int binarySearch(ArrayReader reader, int low, int high, int target){
        while(low<=high){
            int mid = low + (high - low)/2;
            if(reader.get(mid) == target){
                return mid;
            }else if(reader.get(mid) > target){
                high = mid - 1;
            }else{
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: Problem-7.py
"""
Time complexity: O(logn)
Space complexity: O(1)
Approach: keep low = 2 and high = low ^ 2. In a loop, while low <= high, Calculate mid.
            if mid = target, return mid
            Everytime mid is less than target,
            if high is not out of bound, update low = mid + 1 and high = low ^ 2
            if mid is greater than target or mid is out of bounds,
             high = mid - 1

"""

class Solution:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        if reader.get(0) == target:
            return 0
        if reader.get(1) == target:
            return 1
        if reader.get(2) == target:
            return 2
        else:
            low = 2
            high = low ** 2
            while (low <= high):

                mid = low + (high - low) // 2

                if reader.get(mid) == target:
                    return mid
                elif (reader.get(mid) == 2 ** 31 - 1) or reader.get(mid) > target:
                    high = mid - 1
                else:
                    if reader.get(high) < target:
                        low = mid + 1
                        high = low
                    else:
                        low = mid + 1
            return -1